sys/linux: add close_range

JiahengHu opened the pull request:
Added description for close_range which is currently in linux-next
(https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=278a5fbaed89dacd04e9d052f4594ffd0e0585de)

JiahengHu commented:
@morehouse @ebiggers #533

ebiggers commented:
> Added description for close_range which is currently in linux-next
(https://git.kernel.org/pub/scm/linux/kernel/git/next/linux-next.git/commit/?id=278a5fbaed89dacd04e9d052f4594ffd0e0585de)

linux-next.git only keeps 3 months of linux-next versions, so this will become a dead link in 3 months.  Either link to linux-next-history.git instead, or link to the latest patch on lore.kernel.org like you did for process_madvise() and fsinfo().

ebiggers commented on sys/linux/sys.txt:
```diff
@@ -494,6 +495,9 @@ pidfd_send_signal(fd fd_pidfd, sig signalno, info ptr[in, siginfo], flags const[
 pidfd_open(pid pid, flags const[0]) fd_pidfd
 pidfd_getfd(pidfd fd_pidfd, fd fd, flags const[0]) fd
 
+close_range(fd fd, max_fd fd, flags const[CLOSE_RANGE_UNSHARE])
```
The actual code considers `CLOSE_RANGE_UNSHARE` to be an optional flag, not required.  So it shouldn't be a `const`.

JiahengHu replied:
I've just changed it to a flag. But from the source code it seems that any flag other than a CLOSE_RANGE_UNSHARE would result in -EINVAL. Why do we still need it to be a flag?

morehouse replied:
It can be 0 and not result in `EINVAL`, right?  Whereas a const implies the flag always must be specified.

ebiggers replied:
Now it's `flags flags[close_range_flags]`, which looks correct.  @JiahengHu please feel free to resolve comment threads yourself -- thanks!

morehouse commented on sys/linux/sys.txt:
```diff
@@ -494,6 +495,9 @@ pidfd_send_signal(fd fd_pidfd, sig signalno, info ptr[in, siginfo], flags const[
 pidfd_getfd(pidfd fd_pidfd, fd fd, flags const[0]) fd
 
+close_range(fd fd, max_fd fd, flags const[CLOSE_RANGE_UNSHARE])
+close_range$MAXFD(fd fd, max_fd const[INT_MAX], flags const[CLOSE_RANGE_UNSHARE])
```
I'm not sure how helpful this special case is.  IIRC syzkaller has heuristics to try random integers occasionally, so we should probably just rely on those for this corner case.

dvyukov replied:
Yes, it should try this without special descriptions.

JiahengHu replied:
The reason I had this special case is because according to the documentation. "INT_MAX should be passed in most of the time", and I feel like trying random integers occasionally wouldn't give us that frequency we desire. Should I just delete this special case?

dvyukov replied:
Is it accepted as a signed or unsigned integer in the kernel? syzkaller will pass -1 for fd from time to time, and if we declare it as `fd[opt]` it will pass -1 more. And if it's declared as unsigned, then -1 will turn into a very large int. Which can do what we want.

I would either delete this, or use a union to specialize this particular arg rather than the whole syscall:
```
fd_or_int_max [
	fd	fd
	max	const[INT_MAX, int32]
]
```

I would assume that syzkaller will be able to test all possible combinations (closing some/none/all/etc fds) without any particular issues without this specialization b/c this is quite local to this single syscall and just requires passing a particular value in a single argument. It's harder to test cases where multiples values need to be in particular values (exponential explosion).
But this is not exact science. There is no single right answer to everything.

morehouse replied:
I think we should delete it.  All it does is increase the frequency we pass `INT_MAX`, which it's hard to say whether more bugs or coverage will be found or not.  Let's keep things simpler unless we have evidence that the frequency needs to be adjusted.

dvyukov replied:
If Matt also says this, let's go for the simpler descriptions.

JiahengHu replied:
I see. Thanks for the explanation, I've just changed this!

ebiggers commented:
>  Added description for close_range which is currently in linux-next (https://lkml.org/lkml/2019/5/24/423).

Now it's an lkml.org link :-(  Please use lore.kernel.org (or equivalently lkml.kernel.org) instead.  The kernel.org mailing list archive works better, is officially maintained by the Linux Foundation, and includes more information about the original emails.

ebiggers commented:
Can you wrap the commit message at 72 columns?  That's pretty standard for git commits.  I hope that soon I won't need to comment about this anymore :-)

JiahengHu commented:
> Can you wrap the commit message at 72 columns? That's pretty standard for git commits. I hope that soon I won't need to comment about this anymore :-)

Sorry I previously completely misunderstood the "wrap the commit message" thing. I've just fixed it. Thanks for the reminder.

ebiggers commented:
> Sorry I previously completely misunderstood the "wrap the commit message" thing. I've just fixed it. Thanks for the reminder.

It's still not quite right since the link is broken between lines.

Links have to be kept intact, i.e. they can exceed the 72-column limit.

It should look like:

    sys/linux: add close_range
  
    Add description for close_range which is currently in linux-next 
    (https://lore.kernel.org/lkml/20200602204219.186620-2-christian.brauner@ubuntu.com/).

(If you're using `vim`, use the `gqip` command.)

JiahengHu commented:
> > Sorry I previously completely misunderstood the "wrap the commit message" thing. I've just fixed it. Thanks for the reminder.
> 
> It's still not quite right since the link is broken between lines.
> 
> Links have to be kept intact, i.e. they can exceed the 72-column limit.
> 
> It should look like:
> 
> ```
> sys/linux: add close_range
> 
> Add description for close_range which is currently in linux-next 
> (https://lore.kernel.org/lkml/20200602204219.186620-2-christian.brauner@ubuntu.com/).
> ```
> 
> (If you're using `vim`, use the `gqip` command.)

I see. Just fixed it.
